### ISSUE TITLE
training scripts for flickr30k

### ISSUE THREAD

**Issue by AllenLiang96** [kdwonn/DivE#9]
In order to reproduce the experiments asap, i use flickr30k datasets as the benchmark and utilize ddp to run the code. 
the modified training script for flickr is as follows:
`CUDA_VISIBLE_DEVICES=0,1,2,3 python3 train.py \
--data_name f30k --cnn_type resnet152 --wemb_type glove \
--margin 0.2 --max_violation --img_num_embeds 4 --txt_num_embeds 4 \
--img_attention --txt_attention --img_finetune --txt_finetune \
--mmd_weight 0.01 --unif_weight 0.01 \
--batch_size 200 --warm_epoch 0 --num_epochs 80 \
--optimizer adamw --lr_scheduler cosine --lr_step_size 30 --lr_step_gamma 0.1 \
--warm_img --finetune_lr_lower 1 \
--lr 1e-3 --txt_lr_scale 1 --img_pie_lr_scale 0.1 --txt_pie_lr_scale 0.1 \
--eval_on_gpu --sync_bn --amp \
--loss smooth_chamfer --eval_similarity smooth_chamfer --temperature 16 \
--txt_pooling rnn --arch slot --txt_attention_input wemb \
--spm_img_pos_enc_type none --spm_txt_pos_enc_type sine \
--spm_1x1 --spm_residual --spm_residual_norm --spm_residual_activation none \
--spm_activation gelu \
--spm_ff_mult 4 --spm_last_ln \
--img_res_pool max --img_res_first_fc \
--spm_input_dim 1024 --spm_query_dim 1024 \
--spm_depth 4 --spm_weight_sharing \
--remark coco_butd_bigru \
--res_only_norm --img_1x1_dropout 0.1 --spm_pre_norm \
--gpo_1x1 --gpo_rnn \
--weight_decay 1e-4 --grad_clip 1 --lr_warmup -1 --unif_residual \
--workers 4 --dropout 0.1 --caption_drop_prob 0.2 --butd_drop_prob 0.2`
But the performance are unsatisfactory, the results are as follows.
<img width="395" alt="image" src="https://github.com/kdwonn/DivE/assets/21366396/72449d15-6fff-46d0-a520-0b1d55cccbd3">

so what's my problem in this trainning script, i will be very grateful if you can point out my mistake, thank you!

**Comment by kdwonn** [comment]
Hello Liang! For the model training on the Flickr30K dataset, please refer to issue #4, where I provided the script I used for Flickr30K. I also observed that setting a high learning rate (1e-3), as in your script, leads to overfitting and unsatisfactory results.

**Comment by AllenLiang96** [comment]
Many thanks for your assistance, i am running the modified code now. and is there a typo for your egg_block.py? In your repository, you write it as agg_block.py, if i have mistaken your understanding, just forget it.

**Comment by kdwonn** [comment]
I haven't knew that there is the typo! 😂 Sorry to bother you, please modify the respective part in the “agg_block.py”.

Warm regards,
Dongwon Kim
Ph.D candidate | Computer Vision Lab | POSTECH
***@***.*** | https://kdwonn.github.io




2023. 11. 30. 23:02, Wenbin Liang ***@***.***> 작성:



Many thanks for your assistance, i am running the modified code now. and is there a typo for your egg_block.py? In your repository, you write it as agg_block.py, if i have mistaken your understanding, just forget it.

**Comment by AllenLiang96** [comment]
hi Dongwon, the given scripts in issue #4 is unavailable for running on flickr30k, because you use a hyperparamether --eval_distance in this bash, but the parameter do not appear in your code  of this repository up to date(2023.12.01).

**Comment by kdwonn** [comment]
That is because the script is from the code I used during research, an un-refactored version. You can ignore the missing parameters. I will update the repository with the cleanly working Flickr30K training script soon.

**Comment by AllenLiang96** [comment]
Hi Dongwan, thanks for your kind reply. Based on the script you provied in issue #4 , the code run successfully. I ran the code with 4xV100 and 8xV100 respectively, the results are as follows. 
<img width="1041" alt="image" src="https://github.com/kdwonn/DivE/assets/21366396/7d82bae7-71b4-484c-b47a-50513aed645d">
<img width="1026" alt="image" src="https://github.com/kdwonn/DivE/assets/21366396/d5b1759d-947b-4b9e-9161-2ee9f7e6655f">
And here is the results of your paper.
<img width="688" alt="image" src="https://github.com/kdwonn/DivE/assets/21366396/e270ca25-bc81-466b-8b06-a596e63af1a8">
The results based on the resnext 101 backbone and bert embedding on flickr30k is 88.8 98.5 99.6 74.3 94.0 96.7, however I only achieve 81.36, 95.07, 97.83 64.87, 87.51, 93.14. So I want to know where I went wrong, could you please give me some advice. The hyperparameters is set by the trainning scripts in issue #4

**Comment by kdwonn** [comment]
Can you give me more information about your experiment? 1) Have you added additional code? 2) Which training method do you use for the multi-GPU training? In the initial issue, you mentioned that you had used `DistributedDataParallel` to run the code - Does that mean you changed the current mult-gpu training code, which is written using `DataParallel`?

**Comment by AllenLiang96** [comment]
Currently i didn't change the code except changing eval_distance to eval_similarity in the trainning scripts . I use ddp because the initial trainning scripts i use do not have multi_gpu arguments, after you give the updated trainning scripts, i leave the code as its orginal version.  To sum, at code level, the code is not modified, at trainning scripts level, the change is mentioned above. I put my training scripts below in case you may want to check. Thanks for your time and attention!
```
python3 train.py \
--data_name f30k --cnn_type resnext101_32x8d --wemb_type glove \
--margin 0.1 --max_violation --img_num_embeds 4 --txt_num_embeds 4 \
--img_attention --txt_attention --img_finetune --txt_finetune \
--mmd_weight 0.01 --unif_weight 0.01 \
--batch_size 64 --warm_epoch 1 --num_epochs 60 \
--optimizer adamw --lr_scheduler step --lr_step_size 20 --lr_step_gamma 0.1 \
--warm_img --finetune_lr_lower 1 \
--lr 1e-5 --txt_lr_scale 10 --img_pie_lr_scale 10 --txt_pie_lr_scale 10 \
--eval_on_gpu --sync_bn --amp \
--loss smooth_chamfer --eval_similarity smooth_chamfer --temperature 16 \
--txt_pooling rnn --arch slot --txt_attention_input wemb \
--spm_img_pos_enc_type sine --spm_txt_pos_enc_type sine \
--spm_1x1 --spm_residual --spm_residual_norm --spm_residual_activation none \
--spm_activation gelu \
--spm_ff_mult 4 --spm_last_ln \
--img_res_pool max --img_res_last_fc --img_res_last_pool \
--spm_input_dim 1024 --spm_query_dim 1024 \
--spm_depth 4 --spm_weight_sharing \
--remark resnext_bert+slot+8-256+f30k+8v100 \
--res_only_norm --spm_pre_norm \
--gpo_aug --lr_warmup 1 --unif_residual \
--workers 32 --dropout 0.1 --crop_size 512 --use_bert --caption_drop_prob 0.2 --grid_drop_prob 0.2 \
--txt_pooling max --weight_decay 1e-2 --multi_gpu --bn_eval --tri_mean_to_max
```
